Dockerfile

```diff
@@ -1,8 +1,59 @@
-ARG VERSION=unspecified
+# Official Docker images are in the form library/<app> while non-official
+# images are in the form <user>/<app>.
+FROM docker.io/library/python:3.12.2-alpine3.19 as compile-stage
 
-FROM python:3.12.0-alpine
+###
+# Unprivileged user variables
+###
+ARG CISA_USER="cisa"
+ENV CISA_HOME="/home/${CISA_USER}"
+ENV VIRTUAL_ENV="${CISA_HOME}/.venv"
 
-ARG VERSION
+# Versions of the Python packages installed directly
+ENV PYTHON_PIP_VERSION=24.0
+ENV PYTHON_PIPENV_VERSION=2023.12.1
+ENV PYTHON_SETUPTOOLS_VERSION=69.1.1
+ENV PYTHON_WHEEL_VERSION=0.42.0
+
+###
+# Install the specified versions of pip, setuptools, and wheel into the system
+# Python environment; install the specified version of pipenv into the system Python
+# environment; set up a Python virtual environment (venv); and install the specified
+# versions of pip, setuptools, and wheel into the venv.
+#
+# Note that we use the --no-cache-dir flag to avoid writing to a local
+# cache.  This results in a smaller final image, at the cost of
+# slightly longer install times.
+###
+RUN python3 -m pip install --no-cache-dir --upgrade \
```

Comment:
I don't know that there's much to be gained from breaking this `RUN` statement up. I think any errors will be fairly obvious because there are only four parts and only the first and last are similar in what's being done. I'm also not convinced there's much to be gained in terms of caching or especially parallelization. Each part _must_ be done in sequence to have an expected result.